### PR TITLE
Handle exception in needs parens with optional chaining

### PR DESCRIPTION
**Pull request by syi0808**
## Description
This PR is just an exception handling that determines that parentheses are necessary when there is tagged template expression and optional chaining.

Fix: #16455

## Checklist

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
**Prettier pr-16500**
[Playground link](https://deploy-preview-16500--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAKA5nGBhaBnCAGzgBkJ0B+AOlQEpaADBkAGhAgAcYBLfZUAIYAnIRADuABWEI8yEAMJiBAT1lsARkIFgA1lgDKAgLaluUOMgBmCvHA1bdBjtrPpkMIQFc7IW0e7uXj5wAB4ccELcJrAKACoRUMLccLJI1oS2bHiuxACKnhDwVjY+AFZ4Ifo5cPmFFmklbACOBfASohyp8ngAtOZwACaDrCAeAtyErrhGRgJyCoQj2VDoxACCMB7c6p5tESRm9emZIAAWMEaEAOqn3PB4zmBw+jJ33ABud8pyYHhqIO9vABJKBDWD6MCRLhrUH6GDKYjFDI+DiiWxXLQcOSolIRd4WNhmWxCGDtAToWZIk7OITEuTqATqOCLNioswwK7cAYwU7IAAcAAY2EI4C1uCKyRS5g1kWwYIzOdzeUgAExsTy2WKM1LHYJGJkDIYDEgCFaeclwABiECEs02rnmuwgIAAvi6gA)

```sh
--parser babel
```

**Input:**

```jsx
(getConsoleLog?.())``
```

**Output:**

```jsx
getConsoleLog?.()``;

```

**Second Output:**

```jsx
SyntaxError: Tagged Template Literals are not allowed in optionalChain. (1:1)
> 1 | getConsoleLog?.()``;
    | ^
  2 |
```

**Comment by syi0808** [comment]
> **Prettier pr-16500** [Playground link](https://deploy-preview-16500--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAKA5nGBhaBnCAGzgBkJ0B+AOlQEpaADBkAGhAgAcYBLfZUAIYAnIRADuABWEI8yEAMJiBAT1lsARkIFgA1lgDKAgLaluUOMgBmCvHA1bdBjtrPpkMIQFc7IW0e7uXj5wAB4ccELcJrAKACoRUMLccLJI1oS2bHiuxACKnhDwVjY+AFZ4Ifo5cPmFFmklbACOBfASohyp8ngAtOZwACaDrCAeAtyErrhGRgJyCoQj2VDoxACCMB7c6p5tESRm9emZIAAWMEaEAOqn3PB4zmBw+jJ33ABud8pyYHhqIO9vABJKBDWD6MCRLhrUH6GDKYjFDI+DiiWxXLQcOSolIRd4WNhmWxCGDtAToWZIk7OITEuTqATqOCLNioswwK7cAYwU7IAAcAAY2EI4C1uCKyRS5g1kWwYIzOdzeUgAExsTy2WKM1LHYJGJkDIYDEgCFaeclwABiECEs02rnmuwgIAAvi6gA)
> 
> ```shell
> --parser babel
> ```
> 
> **Input:**
> 
> ```js
> (getConsoleLog?.())``
> ```
> 
> **Output:**
> 
> ```js
> getConsoleLog?.()``;
> ```
> 
> **Second Output:**
> 
> ```js
> SyntaxError: Tagged Template Literals are not allowed in optionalChain. (1:1)
> > 1 | getConsoleLog?.()``;
>     | ^
>   2 |
> ```

In addition to the case you mentioned, we considered a few more exceptional cases.

File: src/language-js/needs-parens.js [hunk 1]
```diff
@@ -945,6 +945,31 @@ function needsParens(path, options) {
...
+
+      if (
+        parent.type === "TaggedTemplateExpression" &&
+        (node.type === "OptionalMemberExpression" ||
+          node.type === "OptionalCallExpression")
+      ) {
+        return true;
+      }
+
+      if (
```
Comment (fisker) [inline src/language-js/needs-parens.js:957]: I don't think this block is needed, simple `node.type === "ChainExpressin"` should be enough.

Comment (syi0808) [reply src/language-js/needs-parens.js:957]: line 952 for babel parser. And line 957 for another parsers.

#### babel parser structure
스크린샷 2024-07-23 오전 1 47 43

#### acorn parser structure
스크린샷 2024-07-23 오전 1 47 30

Comment (fisker) [reply src/language-js/needs-parens.js:957]: I'm aware the difference, we don't need check children.

Comment (syi0808) [reply src/language-js/needs-parens.js:957]: Babel parser can be noticed by OptionalMemberExpression or OptionalCallExpression at the top if there is an optional chaining.
But for the rest of the parser, i have to explore the optional properties of the child channings. Am I misinterpreting this?

Comment (fisker) [reply src/language-js/needs-parens.js:957]: ChainExpressin always contains optional call or member expression, so we don't need check the descent.

Comment (syi0808) [reply src/language-js/needs-parens.js:957]: 스크린샷 2024-07-23 오후 4 31 53
Logs show that `node.type` is never optional call or member expression.

File: src/language-js/needs-parens.js [hunk 2]
```diff
@@ -945,6 +945,31 @@ function needsParens(path, options) {
...
+        let object = node;
+
+        do {
+          if (object?.optional === true) {
+            return true;
+          }
+
+          object =
+            object.type === "CallExpression" ? object.callee : object.object;
+        } while (object);
```
Comment (fisker) [inline src/language-js/needs-parens.js:970]:
```suggestion
        return true;
```

File: src/language-js/needs-parens.js (hunk 1 above)
Comment (fisker) [reply src/language-js/needs-parens.js:957]: Apply this https://github.com/prettier/prettier/pull/16500/files#r1687578916, you'll see what I mean.

File: src/language-js/needs-parens.js (hunk 2 above)
Comment (syi0808) [reply src/language-js/needs-parens.js:970]: I'm sorry. If there is no optional chaining, it is just member expression, not chain expression. I understood what you said. Thank you for your review.

**Comment by fisker** [comment]
Moved to `shouldAddParenthesesToChainElement` https://github.com/prettier/prettier/pull/16500/commits/ac500788b3d72b28b19c10aa8a8880a4118cbf56

**Comment by fisker** [comment]
There are two other cases need to fix.

```ts
(a?.b)!``;

(a?.b!)``;
```

Can you try to fix them?

**Comment by syi0808** [comment]
> There are two other cases need to fix.
> 
> ```ts
> (a?.b)!``;
> 
> (a?.b!)``;
> ```
> 
> Can you try to fix them?

Okay. I will try this.

**Comment by syi0808** [comment]
@fisker 
I have added exception handling.
But there are many cases (about 3) and it's too complicated to remove the duplication.
Would this implementation be okay now?

File: tests/format/js/tagged-template-expression-ts/format.test.js [hunk 3]
```diff
@@ -0,0 +1 @@
+runFormatTest(import.meta, ["typescript"]);
```
Comment (fisker) [inline tests/format/js/tagged-template-expression-ts/format.test.js:1]: This test need move to the typescript language directory.

File: src/language-js/needs-parens.js [hunk 4]
```diff
@@ -1177,13 +1192,65 @@ Matches following cases:
...
+    ) ||
+    // case (a?.b)!``; in Babel-ts
+    path.match(
+      (node) =>
+        node.type === "OptionalCallExpression" ||
+        node.type === "OptionalMemberExpression",
+      (node, name) =>
+        name === "expression" &&
+        node.type === "TSNonNullExpression" &&
+        node?.extra?.parenthesized !== true,
```
[the remark this answers is not in the document]
Comment (syi0808) [reply src/language-js/needs-parens.js]: Okay, I didn't use this property and processed it to get the same result.

File: tests/format/js/tagged-template-expression-ts/format.test.js (hunk 3 above)
Comment (syi0808) [reply tests/format/js/tagged-template-expression-ts/format.test.js:1]: Done!

File: tests/format/js/tagged-template-expression-ts/non-null-optional-chaning.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+(a?.b!)``;
+
+(a?.b)!``;
```
Comment (fisker) [inline tests/format/js/tagged-template-expression-ts/non-null-optional-chaning.ts:3]: Can you add trailing new line to all the new tests?

Comment (syi0808) [reply tests/format/js/tagged-template-expression-ts/non-null-optional-chaning.ts]: done!

File: tests/format/typescript/tagged-template-expression/non-null-optional-chaning.ts [hunk 6]
```diff
@@ -0,0 +1,3 @@
+(a?.b!)``;
+
+(a?.b)!``;
```
Comment (fisker) [inline tests/format/typescript/tagged-template-expression/non-null-optional-chaning.ts:3]: Missed one

**Comment by fisker** [comment]
Looks good, I'll take another look later.

**Review by fisker — CHANGES_REQUESTED** [review]
Please add a changelog entry https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md#pull-requests

**Comment by syi0808** [comment]
> Please add a changelog entry https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md#pull-requests

I added changelog. But what should I do about duplication error in lint? Should I combine js and ts changelog?

**Review by fisker — APPROVED** [review]
Good job, thanks.

**Comment by fisker** [comment]
@syi0808 

Looks like I found 4 cases that failed

```
((a?.b!)())``;
((a?.b)!())``;
((a?.())!())``;
((a?.()!)())``;
```

**Comment by fisker** [comment]
But should not related to tagged template literals.

**Prettier pr-16500**
[Playground link](https://deploy-preview-16500--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAKAhgfgHQCMCEAlKoSADQgQAOMAltAM7KjoBOrEA7gApsJMp0AG07oAnkwq5W6MAGs4MAMroAtnAAytKHGQAzYQzhSZ8xUqqztAc2QxWAV2Mgjq2ncfO4ADypxWtOqwwgAq-lBstHACBkJGFAw2QnAAig4Q8PqGzgBWDN5KSanpmUix8SAAjiVw3BxUAiDoDAC0OnAAJp3kIPbotEI2AMIQqqroyE1CQj2JUNbJAIIw9rS4DvDc-lo6WXHOABYwqkIA6ge08AyWYHBK-Je0AG6XYpNgDJIgT04AklBdWBKMABGiLAFKGBiZJ7CpUDhGU4yKiTeHRfxPXQUbRGVgwOroazjWHOSysXGTXDoXBwIQtGBfeHaGCnWgdGAHZAADgADBRWHBqrQBQSiRMytkKDBqaz2ZykAAmCgOIwhakxSUgOCqGkdLodDToeYOQlwABiEFY4xWNkm6A2EBAAF8nUA)

```sh
--parser babel-ts
```

**Input:**

```tsx
(a?.b!)()
```

**Output:**

```tsx
a?.b!();

```

This is wrong, it changed the call to optional chainning.

But we format 

```js
(a?.b!).c
```

correctly, so we missed some checks for babel parser, can you try to fix?

**Comment by fisker** [comment]
Maybe it's here https://github.com/prettier/prettier/blob/11c3243463a8c9c62118473ce3a78ba2229047d7/src/language-js/needs-parens.js#L1196-L1198

**Comment by fisker** [comment]
Also, can you move these tests to `chain-expression`? Because I feel we are doing things wrong in for the `NewExpressions` too, we need the same tests for `NewExpressions`, after this get merged.

**Comment by syi0808** [comment]
> correctly, so we missed some checks for babel parser, can you try to fix? (Can be separate PR, since it's different case)

Separating PR may reduce the scope of work and make it easier to manage, but this actually seems to be handling with the exception of chain expression not only related to tagged template literals. I'll fix it soon.

> Also, can you move these tests to chain-expression? Because I feel we are doing things wrong in for the NewExpressions too, we need the same tests for NewExpressions, after this get merged.

I'll move test. and we can also move new expression also into chain expression in this pr?

**Comment by syi0808** [comment]
> But should not related to tagged template literals.
> 
> **Prettier pr-16500** [Playground link](https://deploy-preview-16500--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAKAhgfgHQCMCEAlKoSADQgQAOMAltAM7KjoBOrEA7gApsJMp0AG07oAnkwq5W6MAGs4MAMroAtnAAytKHGQAzYQzhSZ8xUqqztAc2QxWAV2Mgjq2ncfO4ADypxWtOqwwgAq-lBstHACBkJGFAw2QnAAig4Q8PqGzgBWDN5KSanpmUix8SAAjiVw3BxUAiDoDAC0OnAAJp3kIPbotEI2AMIQqqroyE1CQj2JUNbJAIIw9rS4DvDc-lo6WXHOABYwqkIA6ge08AyWYHBK-Je0AG6XYpNgDJIgT04AklBdWBKMABGiLAFKGBiZJ7CpUDhGU4yKiTeHRfxPXQUbRGVgwOroazjWHOSysXGTXDoXBwIQtGBfeHaGCnWgdGAHZAADgADBRWHBqrQBQSiRMytkKDBqaz2ZykAAmCgOIwhakxSUgOCqGkdLodDToeYOQlwABiEFY4xWNkm6A2EBAAF8nUA)
> 
> ```shell
> --parser babel-ts
> ```
> 
> **Input:**
> 
> ```tsx
> (a?.b!)()
> ```
> 
> **Output:**
> 
> ```tsx
> a?.b!();
> ```
> 
> This is wrong, it changed the call to optional chainning.
> 
> But we format
> 
> ```js
> (a?.b!).c
> ```
> 
> correctly, so we missed some checks for babel parser, can you try to fix? (Can be separate PR, since it's different case)

I modified existing implementation it to cover this.

File: tests/format/typescript/temp/temp.ts [hunk 7]
```diff
@@ -0,0 +1 @@
+a(b?.c!);
```
Comment (fisker) [inline tests/format/typescript/temp/temp.ts:1]: I guess you added this to debug? you can do `yarn debug test.ts`

Comment (syi0808) [reply tests/format/typescript/temp/temp.ts:1]: Sorry i will remove it. And thanks for sharing cool command.

**Comment by fisker** [comment]
> we can also move new expression also into chain expression in this pr?

Let's not make this huge. We don't know what we will face. I'm going to merge this first.

**Comment by fisker** [comment]
Would like to continue work on this?

In next PR, we duplicate tests in this PR. And change


For `NewExpression`

```diff
- ... ``;
+ new ...;
```


For `MemberExpression` (second one make sure we checking .object not .property)

```diff
- ... ``;
+ ... . foo;
+ foo [...];
```


For `CallExpression` (second one make sure we checking .callee not .arguments)

```diff
- ... ``;
+ ... (foo);
+ foo(...);
```

**Review by fisker — APPROVED** [review]

**Comment by syi0808** [comment]
> Would like to continue work on this?
> 
> In next PR, we duplicate tests in this PR. And change
> 
> For `NewExpression`
> 
> ```diff
> - ... ``;
> + new ...;
> ```
> 
> For `MemberExpression` (second one make sure we checking .object not .property)
> 
> ```diff
> - ... ``;
> + ... . foo;
> + foo [...];
> ```
> 
> For `CallExpression` (second one make sure we checking .callee not .arguments)
> 
> ```diff
> - ... ``;
> + ... (foo);
> + foo(...);
> ```

Thank you for organizing. I'll research this and submit PR soon.

**Comment by fisker** [comment]
After we add/fix all tests.

We need rewrite the existing logic, currently it's a mass.

The logic should be:

1. Check if it's root element of ChainExpression (Need two version, one for babel, one for estree)
2. Skip `TSNonNullExpession`
3. Check if it's `TemplateLiteral.tag` / `NewExpression.callee` / `CallExpression.callee` / `MemberExpression.object`

This can be done in the followup PR.

**Pull request merged**